stonesoup/updater/recursive.py

```diff
@@ -0,0 +1,426 @@
+        a predicted state or predicted measurement and an actual measurement,
+        calculate the posterior state.
+        Parameters
+        ----------
+        hypothesis : :class:`~.SingleHypothesis`
+            may carry a predicted measurement, or a predicted state. In the
+            latter case a predicted measurement will be calculated.
+        **kwargs : various
+            These are passed to :meth:`predict_measurement`
+        Returns
+        -------
+        : :class:`~.GaussianStateUpdate`
+            The posterior state Gaussian with mean :math:`\mathbf{x}_{k|k}` and
+            covariance :math:`P_{x|x}`
+        """
+        # Get the predicted state out of the hypothesis
+        predicted_state = hypothesis.prediction
+
+        # Get the measurement model out of the measurement if it's there.
+        # If not, use the one native to the updater (which might still be
+        # none)
+        measurement_model = hypothesis.measurement.measurement_model
+        measurement_model = self._check_measurement_model(measurement_model)
+
+        # If there is no measurement prediction in the hypothesis then do the
+        # measurement prediction (and attach it back to the hypothesis).
+        if hypothesis.measurement_prediction is None:
+            # Attach the measurement prediction to the hypothesis
+            hypothesis.measurement_prediction = self.predict_measurement(
+                predicted_state, measurement_model=measurement_model, **kwargs)
+
+        if not self.number_steps > 0:
+            raise ValueError("Updater cannot run 0 times (or less). This would not provide an "
+                             "updated state")
+
+        nhypothesis = copy.copy(hypothesis)
+        for _ in range(self.number_steps):
+
+            nhypothesis.measurement_prediction = self.predict_measurement(
+                nhypothesis.prediction, measurement_model=measurement_model, n=self.number_steps)
+            # Kalman gain and posterior covariance
+            posterior_covariance, kalman_gain = self._posterior_covariance(nhypothesis)
+
+            # Posterior mean
+            posterior_mean = self._posterior_mean(nhypothesis.prediction, kalman_gain,
+                                                  nhypothesis.measurement,
+                                                  nhypothesis.measurement_prediction)
+            nhypothesis.prediction = Prediction.from_state(
+                nhypothesis.prediction, state_vector=posterior_mean, covar=posterior_covariance)
+
+        if self.force_symmetric_covariance:
+            posterior_covariance = \
+                (posterior_covariance + posterior_covariance.T)/2
+
+        return Update.from_state(
+            hypothesis.prediction,
+            posterior_mean, posterior_covariance,
+            timestamp=hypothesis.measurement.timestamp, hypothesis=hypothesis)
+
+
+class JosephCovarianceRecursiveUpdater(ExtendedKalmanUpdater):
```

Comment:
Could this just inherit from `BayesianRecursiveUpdater`. Just update relevant methods to take `**kwargs`. I guess you could even have `JosephCovariance` as a flag and if statement; and this could even be added to regular KF as well.

Reply:
Added use_jacobian_cov flag to `BayesianRecursiveUpdater`, and if statement to _posterior_covariance() method that will use the alternate calculation dependent on the flag. Removed `JosephCovarianceRecursiveUpdater` as no longer needed.